qualang_tools/addons/calibration/calibrations.py

```diff
@@ -0,0 +1,1208 @@
+                        "Amplitude scan is not available for Ramsey measurements."
+                    )
+                elif (
+                    current_calib == "Resonator_spec"
+                ) and self.flux_line_elmt is None:
+                    raise Exception(
+                        "Amplitude scan is not available for resonator spectroscopy without flux line."
+                    )
+                if var1 is None:
+                    var1 = declare(fixed)
+                    var1_scan = self.scan_var[current_calib][var_name]
+                    var1_name = var_name
+                else:
+                    var2 = declare(fixed)
+                    var2_scan = self.scan_var[current_calib][var_name]
+                    var2_name = var_name
+            elif var_name == "duration":
+                if (self.scan_var[current_calib][var_name] < 4).any():
+                    raise Exception(
+                        "No scanned duration can be lower than 4 clock cycles."
+                    )
+                if current_calib in ["Rabi", "Ramsey"]:
+                    pulse = self.config["elements"][self.qubit_elmt]["operations"][
+                        self.qubit_op
+                    ]
+                    for inpt in self.config["pulses"][pulse]["waveforms"]:
+                        wf = self.config["pulses"][pulse]["waveforms"][inpt]
+                        if self.config["waveforms"][wf]["type"] == "arbitrary":
+                            if (
+                                self.scan_var[current_calib][var_name]
+                                < self.config["pulses"][pulse]["length"] // 4
+                            ).any():
+                                raise Exception(
+                                    "The scanned duration must be greater than the one defined in the configuration for arbitrary waveforms. Arbitrary waveforms can only be stretched and not compressed on the fly."
+                                )
+                if var1 is None:
+                    var1 = declare(int)
+                    var1_scan = self.scan_var[current_calib][var_name]
+                    var1_name = var_name
+                else:
+                    var2 = declare(int)
+                    var2_scan = self.scan_var[current_calib][var_name]
+                    var2_name = var_name
+
+            if len(self.scan_var[current_calib].keys()) == 2:
+                return n, var1, var1_scan, var1_name
+        return n, var1, var1_scan, var1_name, var2, var2_scan, var2_name
+
+    def _align_measure_save(self, I, Q, I_st, Q_st, current_calib, align_flag=True):
+        if align_flag:
+            align(self.qubit_elmt, self.readout_elmt)
+        # Measure the resonator
+        if len(self.out) == 2:
+            measure(
+                self.readout_op,
+                self.readout_elmt,
+                None,
+                dual_demod.full(self.iw[0], self.out[0], self.iw[1], self.out[1], I),
```

Comment:
Can this be used when the upconversion does LO - IF? then the integration weights need to be flipped around. or do we assume that we would swap the cables in the OPX or in the config?

Reply:
If you put a positive IF and you get LO-IF, then you need to flip things. I think that this is the best practice and we should strongly encourage our customers to do so.
It could also be that you output LO+IF but you still need to flip the integration weights, this is if the down conversation cables are swapped... In this case, we also need to flip it (config or physically).
It also drastically simplifies the code, and specifically things like the discriminators and weight optimizers.

Reply:
I agree with what Yoav said and here you can anyway just swap the integration weights that you input to the function when you declare the calibration API (cf line 164 of calibration_test.py) so it should not change this function